jgrapht-core/src/main/java/org/jgrapht/alg/scoring/ClosenessCentrality.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * (C) Copyright 2017-2017, by Dimitrios Michail and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.scoring;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+import org.jgrapht.DirectedGraph;
+import org.jgrapht.Graph;
+import org.jgrapht.alg.interfaces.ShortestPathAlgorithm;
+import org.jgrapht.alg.interfaces.ShortestPathAlgorithm.SingleSourcePaths;
+import org.jgrapht.alg.interfaces.VertexScoringAlgorithm;
+import org.jgrapht.alg.shortestpath.DijkstraShortestPath;
+import org.jgrapht.alg.shortestpath.FloydWarshallShortestPaths;
+import org.jgrapht.graph.EdgeReversedGraph;
+
+/**
+ * Closeness centrality. Computes the closeness centrality of each vertex of a graph. See
+ * <a href="https://en.wikipedia.org/wiki/Closeness_centrality">wikipedia</a> for the definition of
+ * closeness centrality.
+ *
+ * <p>
+ * This implementation computes by default the closeness centrality using outgoing paths and
+ * normalizes the scores. This behavior can be adjusted by the constructor arguments.
+ * 
+ * <p>
+ * Note that the closeness centrality is not defined when the graph is disconnected. Nevertheless,
+ * if there is no (directed) path between two vertices then the total number of vertices is used in
```

Comment:
That part of the wiki description refers to the harmonic centrality which is a variation in order to handle disconnected graphs. I added a separate class called `HarmonicCentrality` to handle disconnected graphs. 

The original approach of adding the number of vertices was proposed by the authors of igraph. It tries to use a number larger than the diameter (at least for the unweighted case). 

Since we also have the harmonic variant, I reverted to the definition, returning 0 for all vertices in case of disconnected graphs.